TimePicker moves to minute mode after hour selection

darrenaustin opened the pull request:
## Description

Adds a feature of the native Android Time Picker to our Material Time Picker. When the user selects an hour, it automatically switches to minute mode. 

This is a merging of two pull requests:
- Code changes from @sdolski https://github.com/flutter/flutter/pull/24677
- Tests from @lucaslcode https://github.com/flutter/flutter/pull/29876

Thanks to both of you for your contributions!

## Related Issues

Fixes #17276, #29600

## Tests

The following tests were added from @lucaslcode:

'tap-select switches from hour to minute'
'drag-select switches from hour to minute'

to time_picker_test.dart.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

darrenaustin committed:
TimePicker moves to minute mode after hour selection

This is a merging of two pull requests:
- Code changes from @sdolski
- Tests from @lucaslcode

darrenaustin committed:
Changed to a more descriptive VoidCallback instead of just a Function for the onHourSelected callback.

HansMuller commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1000,12 +1000,14 @@ class _Dial extends StatefulWidget {
     @required this.mode,
     @required this.use24HourDials,
     @required this.onChanged,
+    @required this.onHourSelected,
```
Even though this is private etc, It looks like the implementation assumes that onHourSelected is non-null; should assert as much here. If there similar assumptions about any of the other parameters, maybe more asserts too.

HansMuller commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -134,6 +134,36 @@ void _tests() {
+    TimeOfDay result;
+
+    final Offset center = await startPicker(tester, (TimeOfDay time) { result = time; });
+    await tester.tapAt(Offset(center.dx, center.dy + 50.0));
```
It would be nice to define variables for the offsets, like hour6, like in the other tests.

HansMuller reviewed (approved):
LGTM with some trivial changes

darrenaustin committed:
Updates from PR review:
- Added more asserts for the _Dial constructor
- Null guarded the onHourSelected callback
- Extracted some variables in the test to make it more explicit what it was doing.

up3 commented:
How do I set it to no show the minutes (like t was before) ?
